File: day2.py
import re

with open("day2.txt", "r") as f:
    text = f.readlines()

pattern = re.compile("([0-9]+)-([0-9]+) (.): (.+)")

count = 0
for line in text:
    start, end, letter, pw = re.search(pattern, line).groups()
    letter_count = pw.count(letter)
    if letter_count >= int(start) and letter_count <= int(end):
        count += 1
print(count)


count = 0
for line in text:
    start, end, letter, pw = re.search(pattern, line).groups()
    if (pw[int(start) - 1] == letter) ^ (pw[int(end) - 1] == letter):
        count += 1
print(count)

File: day2.rb
counts = [0,0]
File.read("day2.txt").split("\n").each do |line|
    from, to, letter, pw = line.match(/(\d+)-(\d+) (\w): (\w+)/i).captures
    letter_count = pw.count letter
    
    if letter_count >= from.to_i and letter_count <= to.to_i then
        counts[0] += 1
    end

    if (pw[from.to_i - 1] == letter) ^ (pw[to.to_i - 1] == letter) then
        counts[1] += 1
    end
end
puts counts